subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptHandler.java

```diff
@@ -139,6 +140,11 @@ private void defineConfiguration() {
             attributes.attribute(Bundling.BUNDLING_ATTRIBUTE, instantiator.named(Bundling.class, Bundling.EXTERNAL));
             attributes.attribute(TargetJvmVersion.TARGET_JVM_VERSION_ATTRIBUTE, Integer.parseInt(JavaVersion.current().getMajorVersion()));
             attributes.attribute(GradlePluginApiVersion.GRADLE_PLUGIN_API_VERSION_ATTRIBUTE, instantiator.named(GradlePluginApiVersion.class, GradleVersion.current().getVersion()));
+
+            classpathConfiguration.getDependencyConstraints().add(dependencyHandler.getConstraints().create(Log4jBannedVersion.LOG4J2_CORE_COORDINATES, constraint -> constraint.version(version -> {
+                version.strictly(Log4jBannedVersion.LOG4J2_CORE_STRICT_VERSION_RANGE);
+                version.prefer(Log4jBannedVersion.LOG4J2_CORE_PREFERRED_VERSION);
+            })));
```

Comment:
Might it be worth forcing an upgrade on API as well? Under this system, you get the following dependencies if `strictly` was used on API (one of my projects had this):
```
runtimeClasspath - Runtime classpath of source set 'main'.
+--- org.apache.logging.log4j:log4j-api:{strictly 2.14.1} -> 2.14.1
+--- org.apache.logging.log4j:log4j-core:2.14.1 -> 2.15.0
|    \--- org.apache.logging.log4j:log4j-api:2.15.0 -> 2.14.1
\--- org.apache.logging.log4j:log4j-core:{strictly [2.15, 3[; prefer 2.15.0} -> 2.15.0 (c)
```

IMO this resolution is already a little weird because it allowed `2.15.0` to be downgraded to `2.14.1,` which to my understanding should not occur under a `require`; but ignoring that: this causes runtime errors. Including this as a default constraint might be surprising for those using strictly on their log4j-api dependency, as it will only fail at runtime!

Reply:
Thanks for sharing, and yes a first level `strictly` will downgrade a second level `require`, that is the whole point of feature.
I wonder if we should bother. Unfortunately, because Log4j does not publish Gradle Module Metadata, it does not benefit from alignment. Though even if it did, this would cause the `strictly` to conflict and you would have to add something to the build script.

Note that we are revising the way this constraint is expressed to be in the shape of an `reject` which is more correct in the long term.